# Week 2 - Practice Questions

## Classification Models

### Question 1: Train a Logistic Regression Model on Titanic Dataset

In [24]:
import pandas as pd

In [25]:
df = pd.read_csv("../Week-1/Titanic_Mini_Project/cleaned_titanic.csv")

df.head()

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked_C,Embarked_Q,Embarked_S
0,1,0,3,"Braund, Mr. Owen Harris",1,22.0,1,0,A/5 21171,7.2500,Unknown,0.0,0.0,1.0
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",0,38.0,1,0,PC 17599,71.2833,C85,1.0,0.0,0.0
2,3,1,3,"Heikkinen, Miss. Laina",0,26.0,0,0,STON/O2. 3101282,7.9250,Unknown,0.0,0.0,1.0
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",0,35.0,1,0,113803,53.1000,C123,0.0,0.0,1.0
4,5,0,3,"Allen, Mr. William Henry",1,35.0,0,0,373450,8.0500,Unknown,0.0,0.0,1.0


In [26]:
X = df[["Pclass", "Sex", "Age", "SibSp", "Parch", "Fare",
        "Embarked_C", "Embarked_Q", "Embarked_S"]]

y = df["Survived"]

In [27]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42
)

print("Training data:", X_train.shape)
print("Testing data:", X_test.shape)

Training data: (712, 9)
Testing data: (179, 9)


In [28]:
from sklearn.linear_model import LogisticRegression

In [29]:
model = LogisticRegression(max_iter=1000)

In [30]:
model.fit(X_train, y_train)

LogisticRegression(max_iter=1000)

## Question 2: Create Confusion Matrix and Accuracy Score

In [31]:
from sklearn.metrics import accuracy_score, confusion_matrix

y_pred = model.predict(X_test)

accuracy = accuracy_score(y_test, y_pred)
cm = confusion_matrix(y_test, y_pred)

print("Accuracy:", accuracy)
print("Confusion Matrix:")
print(cm)

Accuracy: 0.8100558659217877
Confusion Matrix:
[[90 15]
 [19 55]]


## Question 3: Compare Decision Tree and KNN Accuracy

In [32]:
from sklearn.tree import DecisionTreeClassifier

dt_model = DecisionTreeClassifier(random_state=42)

dt_model.fit(X_train, y_train)

dt_pred = dt_model.predict(X_test)

dt_accuracy = accuracy_score(y_test, dt_pred)

print("Decision Tree Accuracy:", dt_accuracy)

Decision Tree Accuracy: 0.7877094972067039


In [33]:
from sklearn.neighbors import KNeighborsClassifier

knn_model = KNeighborsClassifier()

knn_model.fit(X_train, y_train)

knn_pred = knn_model.predict(X_test)

knn_accuracy = accuracy_score(y_test, knn_pred)

print("KNN Accuracy:", knn_accuracy)

KNN Accuracy: 0.7150837988826816


In [34]:
print("Decision Tree Accuracy:", dt_accuracy)
print("KNN Accuracy:", knn_accuracy)

if dt_accuracy > knn_accuracy:
    print("Decision Tree performed better than KNN.")
elif knn_accuracy > dt_accuracy:
    print("KNN performed better than Decision Tree.")
else:
    print("Both models have the same accuracy.")

Decision Tree Accuracy: 0.7877094972067039
KNN Accuracy: 0.7150837988826816
Decision Tree performed better than KNN.


## Question 4: Tune max_depth of Decision Tree

In [35]:
depths = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]

for depth in depths:
    model = DecisionTreeClassifier(max_depth=depth, random_state=42)
    model.fit(X_train, y_train)

    pred = model.predict(X_test)
    accuracy = accuracy_score(y_test, pred)

    print("max_depth:", depth, "Accuracy:", accuracy)

max_depth: 1 Accuracy: 0.7821229050279329
max_depth: 2 Accuracy: 0.7653631284916201
max_depth: 3 Accuracy: 0.7988826815642458
max_depth: 4 Accuracy: 0.7988826815642458
max_depth: 5 Accuracy: 0.7988826815642458
max_depth: 6 Accuracy: 0.7988826815642458
max_depth: 7 Accuracy: 0.7988826815642458
max_depth: 8 Accuracy: 0.7821229050279329
max_depth: 9 Accuracy: 0.7988826815642458
max_depth: 10 Accuracy: 0.7932960893854749


## Conclusion

The Decision Tree model was tested with different max_depth values.

The highest test accuracy was 79.89%, achieved at max_depth values 3, 4, 5, 6, 7, and 9.

Changing max_depth affected the model accuracy, but the highest accuracy was the same for several depth values.